core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -283,32 +284,55 @@ public Snapshot apply() {
       throw new RuntimeIOException(e, "Failed to write manifest list file");
     }
 
+    Map<String, String> summary = summary();
+    String operation = operation();
+
     Long addedRows = null;
-    Long lastRowId = null;
-    if (base.rowLineageEnabled()) {
-      addedRows = calculateAddedRows(manifests);
-      lastRowId = base.nextRowId();
+    Long firstRowId = null;
+    if (base.formatVersion() >= 3) {
+      addedRows = calculateAddedRows(operation, summary, manifests);
+      firstRowId = base.nextRowId();
     }
 
     return new BaseSnapshot(
         sequenceNumber,
         snapshotId(),
         parentSnapshotId,
         System.currentTimeMillis(),
-        operation(),
-        summary(base),
+        operation,
+        summaryWithTotals(base, summary),
         base.currentSchemaId(),
         manifestList.location(),
-        lastRowId,
+        firstRowId,
         addedRows);
   }
 
-  private Long calculateAddedRows(List<ManifestFile> manifests) {
+  private Long calculateAddedRows(
+      String operation, Map<String, String> summary, List<ManifestFile> manifests) {
+    if (summary != null) {
+      long addedRecords =
+          PropertyUtil.propertyAsLong(summary, SnapshotSummary.ADDED_RECORDS_PROP, 0L);
+      if (DataOperations.REPLACE.equals(operation)) {
+        long replacedRecords =
+            PropertyUtil.propertyAsLong(summary, SnapshotSummary.DELETED_RECORDS_PROP, 0L);
+        // added may be less than replaced when records are already deleted by delete files
+        Preconditions.checkArgument(
```

Comment:
Style comment: Does it make sense for this check to be here? Seems like it should be a base check for a Replace Operation and unrelated to calculateAddedRows?